src/dotnet/ShellShimMaker/ShellShimMaker.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.ShellShimMaker
+{
+    public class ShellShimMaker
+    {
+        private readonly string _pathToPlaceShim;
+
+        public ShellShimMaker(string pathToPlaceShim)
+        {
+            _pathToPlaceShim =
+                pathToPlaceShim ?? throw new ArgumentNullException(nameof(pathToPlaceShim));
+        }
+
+        public void CreateShim(string packageExecutablePath, string shellCommandName)
+        {
+            var packageExecutable = new FilePath(packageExecutablePath);
+
+            var script = new StringBuilder();
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                script.AppendLine("@echo off");
+                script.AppendLine($"dotnet {packageExecutable.ToQuotedString()} %*");
+            }
+            else
+            {
+                script.AppendLine("#!/bin/sh");
+                script.AppendLine($"dotnet {packageExecutable.ToQuotedString()} \"$@\"");
+            }
+
+            var scriptPath = GetScriptPath(shellCommandName);
+            try
+            {
+                File.WriteAllText(scriptPath.Value, script.ToString());
+            }
+            catch (UnauthorizedAccessException e)
+            {
+                throw new GracefulException(
```

Comment:
Is this possible? When will the user not have access to the user folder? Also, running as SUDO really solves this? Because then it will place the shim on SUDO's user folder, no? This message may not be right.

Reply:
Yes are right. This should not happen, no try catch here

Reply:
fixed